### DISCUSSION TITLE
Tensor Storage Format

### DISCUSSION THREAD

**Discussion by jab0707** [SCIInstitute/SCIRun#2207] [Q&A] [answered]
Is there anywhere on the documentation that describes the tensor storage format. @dcwhite

**Comment by dcwhite** [comment]
Do you want in-memory storage, or on-disk storage? If the latter, which file type?

**Comment by jab0707** [comment]
What I mean is what are the explicit values you would have assigned to each node of a mesh with tensor data on it. Are the 6 values? 9? What is their order etc.

**Comment by dcwhite** [comment]
The doc for the parser modules has this info, which I think maps onto the in-memory storage scheme:
```
To construct a tensor use the following functions:

 MYTENSOR = tensor(XX,XY,XZ,YY,YZ,ZZ); 

 MYTENSOR = tensor(SCALAR); 

 MYTENSOR = tensor(EIGVEC1,EIGVEC2,EIGVAl1,EIGVAL2,EIGVAL3); 
```

The first constructor with 6 values is what is used in the code as well, but additional data is stored in memory: a 3x3 matrix, and eigen vectors and value data. Any more thoughts @jessdtate @RubioJr9?

**Reply by jab0707** [reply]
Is there a general overview of datatypes and such in the documentation? I tried looking their first for this info.

**Reply by dcwhite** [reply]
This is probably the best available document right now: https://www.sci.utah.edu/devbuilds/scirun_docs/UserGuide.pdf

We really ought to update it and get the content in a more 21st-century format (mesh video with fancy fly-through graphics anyone?).

**Reply by jab0707** [reply]
The thing that I am still confused about is that, when I save out a tensor from SCIRun as a matrix I get 6 values. I assume that these are the xx, xy,xz,yy,yz, and zz values but I do not see anywhere that confirms that this is the correct order.

**Comment by jab0707** [comment, marked as answer]
Ok So I dove into the code and figured it out. Using the tensor() function in the SCIRun API we can create a scirun tensor. There are various inputs but lets say we are using the tensor(vector, vector, scalar, scalar) input. 
```T = tesnor(V1,V2,S1,S2)```
In this case V1 and V2 are column vectors describing the primary and secondary directions of the tensor, while S1 and S2 are the conductivities/scales for those directions. The match that SCIRun does is equivalent to the following:
1) create a third vector V3 such that V3 = cross(V1,V2)
2) create a matrix V = [V1,V2,V3]
3) create a matrix S = [S1, 0, 0; 0, S2, 0; 0, 0, S3] where S3 = S2
4) compute the tensor T as T = M@S@M^T where @ is matrix multiplication and ^T is transpose
The tensor will have entries of:
```
T = [xx xy xz
       yx yy yz
       zx zy zz]
 ```
Scirun then assumes that the tensor is symmetric such that xy = yx, xz = zx, yz = zy. Thus only xx, xy, xz, yy, yz, and zz need to be stored, in that order.

**Reply by tpat95** [reply]
Hi Jake. You are correct about the tensor formatting. The stream operators that convert the data to and from files also use that xx, xy, xz, yy, yz, zz ordering.

Sorry I didn't see this post sooner. My github notifications get buried in my umail account. I switched it to my other email so it doesn't happen again. I'm happy to answer any other questions about tensors in SCIRun!
